keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,645 @@
+	// If this value is nil, the pod's terminationGracePeriodSeconds will be used. Otherwise, this
+	// value overrides the value provided by the pod spec.
+	// Value must be non-negative integer. The value zero indicates stop immediately via
+	// the kill signal (no opportunity to shut down).
+	// This is a beta field and requires enabling ProbeTerminationGracePeriod feature gate.
+	// +optional
+	TerminationGracePeriodSeconds *int64
+}
+```
+
+### Existing Behavior
+
+Taking the Seconds fields, in order of the struct.
+InitialDelaySeconds
+TimeoutSeconds
+PeriodSeconds
+
+#### Defaulting logic
+ - InitialDelaySeconds, has no defaulting logic.
+   Therefore it would get the golang defaulting logic, and default to 0.
+ - TimeoutSeconds, defaults to 1 second if unset (which includes the explicitly set to 0 state). This is the last ending time before the timeout fails. Defaulting, thus, results in slow failures (when one considers a full second of time in cpu time).
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L224-L226
+
+```go
+ 	if obj.TimeoutSeconds == 0 {
+		obj.TimeoutSeconds = 1
+	}
+```
+
+ - Period Seconds is the biggest hurdle in charging and overall cost to the cluster, but also the most useful if set correctly.
+PeriodSeconds defaults to 10 seconds if unset (which includes the explicitly set to 0 state). Fast failure detections are thus blocked with existing defaults.
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L227-L229
+
+```go
+	if obj.PeriodSeconds == 0 {
+		obj.PeriodSeconds = 10
+	}
+```
+
+#### Validation of fields
+
+Must be non-negative, Zero or greater.
+```go
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.InitialDelaySeconds), fldPath.Child("initialDelaySeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.TimeoutSeconds), fldPath.Child("timeoutSeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.PeriodSeconds), fldPath.Child("periodSeconds"))...)
+
+```
+
+#### Fields to Add
+What fields may be necessary to add?
+
+In pkg/apis/core/types.go
+
+```go
+
+type Probe struct {
+    ...
+    ...
+    // How often (in milliseconds) to offset PeriodSeconds when performing the probe.
+	// +optional
+	PeriodMilliseconds *int32
+	// Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+	// +optional
+	InitialDelayMilliseconds *int32
+	// Length of time (in milliseconds) to offset TimeoutMilliseconds before health checking times out.
+	// +optional
+	TimeoutMilliseconds *int32 (*** removed from alpha)
+}
+
+```
+#### Logic for Added Fields
+What is the least logic that could be used?
+
+The combined value of `PeriodSeconds` and `PeriodMilliseconds` must be greater than 200ms and greater than 500ms for exec probe. And after success falls back to original seconds value.
```

Comment:
It is not clear if these offsets will be applied only once or for each iteration in the KEP uptil this line.

Reply:
added to a few other places to make it more clear... clean up the explanations.. thx!

Reply:
Also add up top please.